clouddriver-elasticsearch/src/main/java/com/netflix/spinnaker/clouddriver/elasticsearch/model/ElasticSearchEntityTagsProvider.java

```diff
@@ -0,0 +1,167 @@
+                                       String entityType,
+                                       String idPrefix,
+                                       Map<String, Object> tags,
+                                       int maxResults) {
+    BoolQueryBuilder queryBuilder = QueryBuilders.boolQuery();
+
+    if (cloudProvider != null) {
+      // restrict to a specific clouddriver (optional)
+      queryBuilder = queryBuilder.must(QueryBuilders.matchQuery("entityRef.cloudProvider", cloudProvider));
+    }
+
+    if (idPrefix != null) {
+      // restrict to a specific id prefix (optional)
+      queryBuilder = queryBuilder.must(QueryBuilders.wildcardQuery("id", idPrefix));
+    }
+
+    return search(entityType, applyTagsToBuilder(queryBuilder, tags), maxResults);
+  }
+
+  @Override
+  public Optional<EntityTags> get(String id) {
+    return get(id, Collections.emptyMap());
+  }
+
+  @Override
+  public Optional<EntityTags> get(String id, Map<String, Object> tags) {
+    BoolQueryBuilder queryBuilder = QueryBuilders.boolQuery().must(QueryBuilders.matchQuery("_id", id));
+    List<EntityTags> entityTags = search(null, applyTagsToBuilder(queryBuilder, tags), 1);
+    return entityTags.isEmpty() ? Optional.empty() : Optional.of(entityTags.get(0));
+  }
+
+  @Override
+  public void index(EntityTags entityTags) {
+    try {
+      Index action = new Index.Builder(objectMapper.convertValue(entityTags, Map.class))
+        .index(activeElasticSearchIndex)
+        .type(entityTags.getEntityRef().getEntityType())
+        .id(entityTags.getId())
+        .build();
+
+      JestResult jestResult = jestClient.execute(action);
+      if (!jestResult.isSucceeded()) {
+        throw new ElasticSearchException(
+          format("Failed to index %s, reason: '%s'", entityTags.getId(), jestResult.getErrorMessage())
+        );
+      }
+    } catch (IOException e) {
+      throw new ElasticSearchException(
+        format("Failed to index %s, reason: '%s'", entityTags.getId(), e.getMessage())
+      );
+    }
+  }
+
+  @Override
+  public void verifyIndex(EntityTags entityTags) {
+    OperationPoller.retryWithBackoff(o -> {
+        // verify that the indexed document can be retrieved (accounts for index lag)
+        if (!get(entityTags.getId(), entityTags.getTags()).isPresent()) {
+          throw new ElasticSearchException(format("Failed to index %s, reason: 'no document found with id'", entityTags.getId()));
+        }
+        return true;
+      },
+      1000,
+      5
+    );
+  }
+
+  private BoolQueryBuilder applyTagsToBuilder(BoolQueryBuilder queryBuilder, Map<String, Object> tags) {
+    if (tags == null) {
+      return queryBuilder;
+    }
+
+    for (Map.Entry<String, Object> entry : tags.entrySet()) {
+      // restrict to specific tags (optional)
+      if (entry.getValue().equals("*")) {
+        queryBuilder = queryBuilder.must(QueryBuilders.existsQuery("tags." + entry.getKey()));
```

Comment:
Treating `*` as convention for "don't care what the tag value is, just that the tag exists".